Add support for AppImages

JakubVanek opened the pull request:
This PR proposes to distribute DataPlotter in an additional packaging format: [AppImage](https://en.wikipedia.org/wiki/AppImage). AppImages are mostly self-contained Linux executables - they have most of the libraries included within them.

## Why

I think that AppImages are a good compromise between the initial number of supported distributions and the effort required to achieve it.
 * One AppImage can work across multiple distributions. Based on (limited) testing via Docker, this DataPlotter AppImage works on Arch, Fedora, Debian and Ubuntu (albeit I have only tested them under X11).
 * The AppImage ships Qt libraries inside, which reduces the need for testing against multiple Qt versions (or working around their [bugs](https://bugreports.qt.io/browse/QTBUG-127575)).

To be fair, there are also some disadvantages:
* The AppImages do not integrate with the rest of the system - they are just executable files.
  * They do not have a desktop icon by default, although some workarounds exist (e.g. [appimaged](https://github.com/AppImageCommunity/appimaged)).
  * They do not update like the rest of the system typically does in Linux distros. Neither do the bundled libraries update.
  * The current DataPlotter AppImage is not able to do what https://github.com/jirimaier/DataPlotter/blob/main/deploy_linux/postinst does.

I think that a good overall Linux support strategy could be the following:
 * Provide an AppImage for all distributions as a "reference" / "known working" download for users.
 * Additionally provide classic repositories for Linux distributions via the OpenSUSE Build Service, just like QtRvSim does. 

### Why not Flatpak instead of AppImage?

I am biased by being a long-time user of Ubuntu. Ubuntu does not ship Flatpak by default, it instead uses Snap (although Flatpak can be installed additionally). Conversely, distributions like Fedora ship Flatpak by default and do not ship Snap (but it can also be installed additionally). I see this as a complication that AppImage does not have. 

However, when looked at from the system integration point of view, Flatpak/Snap are indeed superior solutions to AppImage (e.g. they have desktop icons, they have sandboxing and they have reusable core runtimes). Therefore, if the time resources allow it, I think that providing one or both of these would also be a good idea.

## How to use a built AppImage

You just have to download the `.AppImage` file from GitHub, mark it as executable via `chmod +x <file>.AppImage` and then run it.

There is also an additional `.AppImage.zsync` file built in GitHub Actions. I am not entirely sure how it works, but the way I understand it is the following:
 * There are tools such as https://appimage.github.io/AppImageUpdate/ that allow you to update AppImages automatically.
 * With the current AppImage build configuration, the utilities will look for a `.zsync` file within the latest GitHub release of this project.
 * The `.zsync` file probably allows the utility to download only the changed parts of the AppImage.

## How the packaging works

The process relies on two tools: [appimage-builder](https://github.com/AppImageCrafters/appimage-builder) and [appimagetool](https://github.com/AppImage/appimagetool).
 * `appimage-builder` basically downloads a predefined set of Ubuntu packages (i.e. Qt libraries and their dependencies) and extracts them next to a DataPlotter binary.
   * I decided against `linuxdeploy` and `linuxdeployqt` because they seem to offer less control. With `appimage-builder`, I can be sure that the AppImage will include all specified QML modules, even if DataPlotter does not use them out-of-the-box (they would be used only by user-provided scripts).
 * `appimagetool` packages the directory with DataPlotter and libraries into a self-contained executable.
   * `appimage-builder` can do this too, but it uses an older `appimagetool` that has some disadvantages.

## Miscellaneous

Supersedes: https://github.com/jirimaier/DataPlotter/pull/11
AppImage built in my fork: [Actions log](https://github.com/JakubVanek/DataPlotter/actions/runs/13573306489/job/37943460538), [Artifact zip](https://github.com/JakubVanek/DataPlotter/actions/runs/13573306489/artifacts/2665534142)
Cc: @ppisa (I just want to provide an update that something is happening with this).

JakubVanek commented:
On a bit more pessimistic note, AppImages do seem quite hacky to me. They have to somehow mix system libraries (e.g. OpenGL drivers, possibly from Nvidia) with the bundled libraries (e.g. Qt). This seems to work well for DataPlotter, but it does not look super robust, and it can indeed sometimes break (see "Reasons for moving to Flatpak" in [PrusaSlicer](https://github.com/prusa3d/PrusaSlicer/releases/tag/version_2.9.0-alpha1) release notes).

JakubVanek commented:
Hmm, only now I have realized that DataPlotter already had AppImages in the past. In this case, the PR description is explaining things that did not need to be explained; I'm sorry for that.

Did the AppImages work well back then? I'm thinking that there had to be some reason they were abandoned (apart from the ones I already hit).

jirimaier commented:
The reason why I abandoned that was that I never really automatized the process (I did it manually on Linux VM). When I decided to automate the process, I started from scratch and decided that the deb package may be better simply because I prefer the installer way rather than portable. One particular problem with the appimage is that (as far as I know) the app can change anything in the image, so it can't save its settings.

jirimaier commented:
I would agree to merge this PR in the future, but first, we should organize the actions and Python scripts better so as not to duplicate the code.

JakubVanek commented:
If you'd still be interested, I have rebased the code on top of current `main` and reorganized it to reduce duplication.

JakubVanek commented:
> Hmm, the definition may need some updates, there are now some required libraries that are missing in Fedora 41.

It seems that it is okay after all. The following stuff is needed for smooth operation on Fedora
```
dnf install libGL libEGL libSM libICE harfbuzz fontconfig dejavu-fonts-all fusermount
```
Most of these should be available on desktops. `fusermount` is somewhat optional - the package will work even without it.

ppisa commented:
I am not sure which way to AppImage is the right one. The 
```
make open_build_service_bundle
```
already generates `PKGBUILD` and build on OpenSUSE build service attempts to use it 

https://build.opensuse.org/package/show/home:ppisa/data-plotter

It fails and I am not sure if that is caused by mistake in `PKGBUILD` or by longer time failure of AppImage builds on OBS. It can be probably the second problem because I ma sure that it worked in some past. We can contact SUSE experts...

QtRvSim provides example how to package Qt application to Flatpak 

https://flathub.org/apps/cz.cvut.edu.comparch.qtrvsim

and David Heidelberg  helped us with  Flatpak.

JakubVanek commented:
I think that building AppImage in GH Actions does have one advantage: the AppImage build is available for every commit, not just for released commits. This might be helpful when testing fixes for issues of third-party users.

However, I do agree that the duplication is unfortunate. Also, the `AppImageBuilder.yml` definition might be more fragile than the `appimage.yml.in` processed by OBS.

JakubVanek commented:
However, I wouldn't mind if my mechanism is removed and the OBS is used for building the AppImages.

JakubVanek commented:
> It fails and I am not sure if that is caused by mistake in `PKGBUILD` or by longer time failure of AppImage builds on OBS. It can be probably the second problem because I ma sure that it worked in some past. We can contact SUSE experts...

I think I found a possible cause. I think that the AppImage repository in your OBS project has to include both the AppImage and some base OS repository as paths. For example, https://build.opensuse.org/repositories/home:hawkeye116477:waterfox has these paths:
![obrazek](https://github.com/user-attachments/assets/1314e5ff-dbb4-46ab-a466-1788b6789a87)
While https://build.opensuse.org/repositories/home:ppisa has only
![obrazek](https://github.com/user-attachments/assets/addf3617-b2da-4af1-87d6-adfd873540cd)
The difference is visible in the project-level meta configuration:
* working appimage:
```xml
  <repository name="AppImage">
    <path project="home:hawkeye116477:waterfox" repository="AlmaLinux_8"/>
    <path project="OBS:AppImage" repository="AppImage.leap_15.4"/>
    <arch>x86_64</arch>
  </repository>
```
* not working appimage:
```xml
  <repository name="AppImage">
    <path project="OBS:AppImage" repository="AppImage"/>
    <arch>x86_64</arch>
  </repository>
```

This seems to agree with https://docs.appimage.org/packaging-guide/hosted-services/opensuse-build-service.html#setting-up-an-appimage-target-for-the-project .

However, I am not sure where this is best configured - if at the project level or at the single package level.

JakubVanek commented:
My guess is that using
```xml
  <repository name="AppImage">
    <path project="openSUSE:Leap:15.6" repository="standard"/>
    <path project="OBS:AppImage" repository="toolchain.15.6"/>
    <arch>x86_64</arch>
  </repository>
```
or
```xml
  <repository name="AppImage">
    <path project="home:ppisa" repository="15.6"/>
    <path project="OBS:AppImage" repository="toolchain.15.6"/>
    <arch>x86_64</arch>
  </repository>
```
should do the trick, but I am not certain

JakubVanek commented:
Hmm, I think that
```xml
  <repository name="AppImage">
    <path project="OBS:AppImage" repository="AppImage.leap_15.4"/>
    <arch>x86_64</arch>
  </repository>
```
should work as well. I only now noticed that `AppImage.leap_15.4` includes the SUSE repositories as well.

ppisa commented:
Yes, it seems that reconfiguration of AppImage to use specific base started the build. I have chosen `OBS:AppImage/AppImage.15.4`. I do not remember that there has been such base selection in the past... May it be they have single base at the time when I have started wit QtMips but as it got too out of date then SUSE added more bases for AppImages...

JakubVanek commented:
I'm sorry, I think I was wrong and `OBS:AppImage/AppImage` was the correct choice after all. The build now passes, but only a RPM is built. I don't really know what is going on; I'm trying to debug this in https://build.opensuse.org/project/show/home:linuxtardis:sdi_linux_pkgs but I'm not succeeding.

JakubVanek commented:
I think I finally have it. In the project meta, these options work for me:
```xml
  <repository name="AppImage">
    <path project="OBS:AppImage" repository="AppImage.leap_15.4"/>
    <arch>x86_64</arch>
  </repository>
```
or
```xml
  <repository name="AppImage">
    <path project="OBS:AppImage" repository="toolchain.15.5"/>
    <path project="openSUSE:Leap:15.5:Update" repository="standard"/>
    <arch>x86_64</arch>
  </repository>
```

Then, I have added a `_service` file to my `data-plotter` repo with the following contents:
```xml
<services>
   <service name="appimage"/>
</services>
```

Finally, I have modified the "Project Config" to be the following:
```
%if "%_repository" == "AppImage"

Type: appimage
Repotype: staticlinks
Patterntype: none

FileProvides: /sbin/mkinitrd mkinitrd

Required: build-pkg2appimage
Conflict: build-pkg2appimage:libudev-mini1
Conflict: build-pkg2appimage:systemd-mini
Required: cmake-full
Conflict: cmake-mini
Ignore: !gpg2:pinentry
Ignore: !systemd:kbd
Ignore: !systemd:kmod
Ignore: !systemd:systemd-presets-branding
Ignore: !systemd:dbus-1
Ignore: !systemd:pam-config
Ignore: !systemd:udev
Ignore: !libgio-2_0-0:dbus-1-x11
Ignore: !logrotate:cron
Ignore: !cracklib:cracklib-dict

# build-compare ignore appimages. Means we would not get maintenance updates
Support: !build-compare

# the checks are usually not valid for AppImage files, disable all of them for now
Support: !post-build-checks
Support: !brp-check-suse
Support: !rpmlint-Factory

%endif
```
(most of the code comes from https://build.opensuse.org/projects/OBS:AppImage:Templates:Leap:15.1/prjconf )

This produces an AppImage that can be executed. However, it seems that the AppImage is missing some QML modules, as DataPlotter prints some warnings to console when it is opened.
